README.md
# progress_bar

Progress bar for dart command-line apps. A port of [node-progress][nodeprogress] to [Dart][darthome].

## Usage

Basic usage:

    var bar = new ProgressBar(' [:bar] :percent :etas ', total: 10);
    var timer = new Timer.periodic(new Duration(seconds: 1), (Timer timer) {
      bar.tick();
      if (bar.complete) {
        timer.cancel();
      }
    });

![basic](https://raw.github.com/jarontai/progress_bar/master/example/progress_bar_basic.gif)

### Options

These are keys in the options you can pass to the progress bar along with
`total` as seen in the example above.

- `total` total number of ticks to complete
- `width` the displayed width of the progress bar defaulting to total
- `complete` completion character defaulting to "="
- `incomplete` incomplete character defaulting to "-"
- `clear` option to clear the bar on completion defaulting to false
- `callback` optional function to call when the progress bar completes

### Tokens

These are tokens you can use in the format of your progress bar.

- `:bar` the progress bar itself
- `:current` current tick number
- `:total` total ticks
- `:elapsed` time elapsed in seconds
- `:percent` completion percentage
- `:eta` estimated completion time in seconds

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/jarontai/progress_bar/issues
[nodeprogress]: https://github.com/tj/node-progress
[darthome]: https://www.dartlang.org/

CHANGELOG.md
# Changelog

## 1.0.0
Support Dart 2 and null safety

## 0.2.0

* Fix compile error

## 0.1.1

* Make options optional.
* Add the optional callback function, which called when the progress bar completes.

## 0.1.0

* Initial version.


test/all_test.dart
// Copyright (c) 2014, <Jaron Tai>. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library progress_bar.test;

import 'package:test/test.dart';
import 'package:progress_bar/progress_bar.dart';

main() {
  group('ProgressBar test - ', () {
    late ProgressBar bar;

    setUp(() {
      bar = ProgressBar(':bar', total: 10);
    });

    test('Intialize Test', () {
      expect(bar.curr, 0);
      expect(bar.complete, false);
      expect(bar.total, 10);
      expect(bar.width, 10);
      expect(bar.clear, false);
      expect(bar.completeChar, '=');
      expect(bar.incompleteChar, '-');
    });
  });
}


lib/src/progress_bar.dart
// Copyright (c) 2014, <Jaron Tai>. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library progress_bar.base;

import 'dart:io';
import 'dart:math';

/**
 * ProgressBar
 */
class ProgressBar {
  String format;
  int total;
  int width;
  bool clear;
  String completeChar;
  String incompleteChar;
  Function? callback;
  int curr = 0;
  bool complete = false;
  late DateTime start;
  String? lastDraw;

  /**
   * Initialize a `ProgressBar` with the given `format` string and the `options` map.
   *
   * Format tokens:
   *
   *   - `:bar` the progress bar itself
   *   - `:current` current tick number
   *   - `:total` total ticks
   *   - `:elapsed` time elapsed in seconds
   *   - `:percent` completion percentage
   *   - `:eta` eta in seconds
   *
   * Options:
   *
   *   - `total` total number of ticks to complete
   *   - `width` the displayed width of the progress bar defaulting to total
   *   - `completeChar` completion character defaulting to "="
   *   - `incompleteChar` incomplete character defaulting to "-"
   *   - `callback` optional function to call when the progress bar completes
   *   - `clear` will clear the progress bar upon termination
   *
   */
  ProgressBar(
    this.format, {
    this.total = 0,
    this.width = 0,
    this.clear = false,
    this.completeChar = '=',
    this.incompleteChar = '-',
    this.callback,
  }) {
    if (width <= 0 && total > 0) {
      width = total;
    }
  }

  /**
   * "tick" the progress bar with optional `len` and optional `tokens`.
   */
  tick({int len: 1, Map<String, String>? tokens}) {
    if (this.curr == 0) {
      this.start = new DateTime.now();
    }

    this.curr += len;
    this.render(tokens);

    // progress complete
    if (this.curr >= this.total) {
      this.complete = true;
      this.terminate();
      if (this.callback != null && this.callback is Function) {
        // Function.apply(this.callback, [this.complete]);
      }
    }
  }

  /**
   * Method to render the progress bar with optional `tokens` to place in the
   * progress bar's `format` field.
   *
   */
  render(Map<String, String>? tokens) {
    if (!stdout.hasTerminal) return;

    var ratio = this.curr / this.total;
    ratio = min(max(ratio, 0.0), 1.0);

    var percent = ratio * 100;
    var elapsed = new DateTime.now().difference(this.start).inMilliseconds;
    var eta = (percent == 100) ? 0 : elapsed * (this.total / this.curr - 1);

    /* populate the bar template with percentages and timestamps */
    var str = this
        .format
        .replaceAll(':current', this.curr.toString())
        .replaceAll(':total', this.total.toString())
        .replaceAll(':elapsed',
            elapsed.isNaN ? '0.0' : (elapsed / 1000).toStringAsFixed(1))
        .replaceAll(
            ':eta',
            (eta.isNaN || !eta.isFinite)
                ? '0.0'
                : (eta / 1000).toStringAsFixed(1))
        .replaceAll(':percent', percent.toStringAsFixed(0) + '%');

    /* compute the available space (non-zero) for the bar */
    var availableSpace =
        max(0, stdout.terminalColumns - str.replaceAll(':bar', '').length);
    var width = min(this.width, availableSpace);

    /* the following assumes the user has one ':bar' token */
    var completeLength = (width * ratio).round();
    var completeStr =
        new List<String>.filled(completeLength, this.completeChar).join();
    var incompleteStr =
        new List<String>.filled(width - completeLength, this.incompleteChar)
            .join();

    /* fill in the actual progress bar */
    str = str.replaceAll(':bar', completeStr + incompleteStr);

    /* replace the extra tokens */
    if (tokens != null && tokens.isNotEmpty) {
      tokens.forEach((key, val) {
        str = str.replaceAll(':' + key, val);
      });
    }

    if (this.lastDraw != str) {
      stdout.writeCharCode(13); // output carriage return
      stdout.write(str);
      this.lastDraw = str;
    }
  }

  /**
   * "update" the progress bar to represent an exact percentage.
   * The ratio (between 0 and 1) specified will be multiplied by `total` and
   * floored, representing the closest available "tick." For example, if a
   * progress bar has a length of 3 and `update(0.5)` is called, the progress
   * will be set to 1.
   *
   * A ratio of 0.5 will attempt to set the progress to halfway.
   *
   */
  update(num ratio) {
    var goal = (ratio * this.total).floor();
    var delta = goal - this.curr;
    this.tick(len: delta);
  }

  /**
   * Terminates a progress bar.
   *
   */
  terminate() {
    if (this.clear) {
      for (int i = 0; i < stdout.terminalColumns; i++) {
        stdout.writeCharCode(8); // output backspace
      }
    } else {
      stdout.writeln();
    }
  }
}


lib/src/progress_bar_base.dart
// Copyright (c) 2014, <your name>. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library progress_bar.base;

import 'dart:io';
import 'dart:math';

/**
 * Initialize a `ProgressBar` with the given `format` string and the map `options`.
 *
 * Options:
 *
 *   - `total` total number of ticks to complete
 *   - `width` the displayed width of the progress bar defaulting to total
 *   - `stream` the output stream defaulting to stderr
 *   - `complete` completion character defaulting to "="
 *   - `incomplete` incomplete character defaulting to "-"
 *   - `callback` optional function to call when the progress bar completes
 *   - `clear` will clear the progress bar upon termination
 *
 * Tokens:
 *
 *   - `:bar` the progress bar itself
 *   - `:current` current tick number
 *   - `:total` total ticks
 *   - `:elapsed` time elapsed in seconds
 *   - `:percent` completion percentage
 *   - `:eta` eta in seconds
 *
 */
class ProgressBar {
  String format = '';
  Map options = {
    'total': 0,
    'width': 0,
    'clear': false,
    'complete': '=',
    'incomplete': '-'
  };
  int curr = 0;
  bool complete = false;
  DateTime start;
  String lastDraw;

  ProgressBar(String format, Map options) {
    this.format = format;
    if (options != null) {
      this.options.keys.forEach((key) {
        if (options[key] != null) {
          this.options[key] = options[key];
        }
      });
      if (this.options['width'] == 0) {
        this.options['width'] = this.options['total'];
      }
    }
  }

  /**
   * "tick" the progress bar with optional `len` and optional `tokens`.
   */
  tick({int len: 1, Map tokens}) {
    if (this.curr == 0) {
      this.start = new DateTime.now();
    }

    this.curr += len;
    this.render(tokens);

    // progress complete
    if (this.curr >= this.options['total']) {
      this.complete = true;
      this.terminate();
      return;
    }
  }

  /**
   * Method to render the progress bar with optional `tokens` to place in the
   * progress bar's `format` field.
   *
   */
  render(Map tokens) {
    if (!stdout.hasTerminal) return;

     var ratio = this.curr / this.options['total'];
     ratio = min(max(ratio, 0), 1);

     var percent = ratio * 100;
     var elapsed = new DateTime.now().difference(this.start).inMilliseconds;
     var eta = (percent == 100) ? 0 : elapsed * (this.options['total'] / this.curr - 1);

     /* populate the bar template with percentages and timestamps */
     var str = this.format
       .replaceAll(':current', this.curr.toString())
       .replaceAll(':total', this.options['total'].toString())
       .replaceAll(':elapsed', elapsed.isNaN ? '0.0' : (elapsed / 1000).toStringAsFixed(1))
       .replaceAll(':eta', (eta.isNaN || !eta.isFinite) ? '0.0' : (eta / 1000).toStringAsFixed(1))
       .replaceAll(':percent', percent.toStringAsFixed(0) + '%');

     /* compute the available space (non-zero) for the bar */
     var availableSpace = max(0, stdout.terminalColumns - str.replaceAll(':bar', '').length);
     var width = min(this.options['width'], availableSpace);

     /* TODO: the following assumes the user has one ':bar' token */
     var incomplete, complete, completeLength;
     completeLength = (width * ratio).round();
     complete = new List<String>.filled(completeLength, this.options['complete']).join();
     incomplete = new List<String>.filled(width - completeLength, this.options['incomplete']).join();

     /* fill in the actual progress bar */
     str = str.replaceAll(':bar', complete + incomplete);

     /* replace the extra tokens */
     if (tokens != null) {
       for (var key in tokens) str = str.replaceAll(':' + key, tokens[key]);
     }

     if (this.lastDraw != str) {
       stdout.writeCharCode(13);
       stdout.write(str);
       this.lastDraw = str;
     }

  }

  /**
   * "update" the progress bar to represent an exact percentage.
   * The ratio (between 0 and 1) specified will be multiplied by `total` and
   * floored, representing the closest available "tick." For example, if a
   * progress bar has a length of 3 and `update(0.5)` is called, the progress
   * will be set to 1.
   *
   * A ratio of 0.5 will attempt to set the progress to halfway.
   *
   * @param {number} ratio The ratio (between 0 and 1 inclusive) to set the
   *   overall completion to.
   * @api public
   */
  update() {
//    var goal = Math.floor(ratio * this.total);
//    var delta = goal - this.curr;
//
//    this.tick(delta, tokens);
    throw new UnimplementedError();
  }

  /**
   * Terminates a progress bar.
   *
   * @api public
   */
  terminate() {
    if (this.options['clear']) {
//      stdout.writeln();
    } else {
//      console.log();
    }
    stdout.writeln();
  }
}